src/redprl/refiner.fun

```diff
@@ -708,32 +754,57 @@ struct
-          | _ >> CJ.EQ_TYPE tys => StepEqType sign tys
-          | _ >> CJ.EQ ((m, n), ty) => StepEq sign ((m, n), ty)
-          | _ >> CJ.SYNTH m => StepSynth sign m
+        (fn _ >> CJ.TRUE (ty, _) => StepTrue sign ty
+          | _ >> CJ.EQ_TYPE (tys, _) => StepEqType sign tys
+          | _ >> CJ.EQ ((m, n), (ty, _)) => StepEq sign ((m, n), ty)
+          | _ >> CJ.SYNTH (m, _) => StepSynth sign m
           | _ >> CJ.PARAM_SUBST _ => Misc.ParamSubst
           | MATCH _ => Misc.MatchOperator
           | MATCH_RECORD _ => Record.MatchRecord orelse_ Computation.MatchRecordHeadExpansion sign then_ Record.MatchRecord)
 
+      fun EqTypeFromHyp alpha jdg =
+        let
+          val (_, H) >> CJ.EQ_TYPE ((a0, b0), k0) = jdg
+          val isUnary = Abt.eq (a0, b0)
+          val isUseful =
+            fn CJ.EQ_TYPE ((a1, b1), k1) =>
+                 Abt.eq (a0, a1) andalso Abt.eq (b0, b1)
+                 andalso K.greatestMeetRight' (SOME k0, SOME k1) <> SOME k0
+             | CJ.EQ (_, (a1, k1)) =>
+                 isUnary andalso Abt.eq (a0, a1)
+                 andalso K.greatestMeetRight' (SOME k0, SOME k1) <> SOME k0
+             | CJ.TRUE (a1, k1) =>
+                 isUnary andalso Abt.eq (a0, a1)
+                 andalso K.greatestMeetRight' (SOME k0, SOME k1) <> SOME k0
+             | _ => false
+        in
+          case Hyps.search H isUseful of
+            SOME (lbl, _) =>
+              ( TypeEquality.FromEqTypeHyp lbl orelse_
+                TypeEquality.FromWfHyp lbl orelse_
+                TypeEquality.FromTrueHyp lbl)
+              alpha jdg
+          | NONE => raise E.error [Fpp.text "Could not find suitable hypothesis"]
+        end
 
-      fun isWfJdg (CJ.TRUE _) = false
-        | isWfJdg _ = true
-
-      structure HypsUtil = TelescopeUtil (Hyps)
-
-      fun FindHyp alpha ((I, H) >> jdg) =
-        if isWfJdg jdg then
-          case HypsUtil.search H (fn jdg' => CJ.eq (jdg, jdg')) of
-             SOME (lbl, _) => Hyp.Project lbl alpha ((I, H) >> jdg)
-           | NONE => raise E.error [Fpp.text "Could not find suitable hypothesis"]
-        else
-          raise E.error [Fpp.text "Non-deterministic tactics can only be run on auxiliary goals"]
+      fun EqFromWfHyp alpha jdg =
+        let
+          val (_, H) >> CJ.EQ ((m0, n0), (a0, k0)) = jdg
+          val isUseful =
+            fn CJ.EQ ((m1, n1), (a1, k1)) =>
+                Abt.eq (m0, m1) andalso Abt.eq (n0, n1) andalso Abt.eq (a0, a1)
+                andalso K.greatestMeetRight' (SOME k0, SOME k1) <> SOME k0
+             | _ => false
+        in
+          case Hyps.search H isUseful of
+            SOME (lbl, _) => Equality.FromWfHyp lbl alpha jdg
+          | NONE => raise E.error [Fpp.text "Could not find suitable hypothesis"]
+        end
     in
       fun AutoStep sign alpha jdg = 
         StepJdg sign alpha jdg
           handle exn => 
-            FindHyp alpha jdg
+            (EqTypeFromHyp orelse_ EqFromWfHyp) alpha jdg
```

Comment:
@jonsterling The actual change is that the above complicated search replaced `Hyp.Project` in `AutoStep`.